[SPARK-27734][CORE][SQL] Add memory based thresholds for shuffle spill

amuraru opened the pull request:
## What changes were proposed in this pull request?

When running large shuffles (700TB input data, 200k map tasks, 50k reducers on a 300 nodes cluster) the job is regularly OOMing in map and reduce phase.

IIUC ShuffleExternalSorter (map side) and ExternalAppendOnlyMap and ExternalSorter (reduce side) are trying to max out the available execution memory. This in turn doesn't play nice with the Garbage Collector and executors are failing with OutOfMemoryError when the memory allocation from these in-memory structure is maxing out the available heap size (in our case we are running with 9 cores/executor, 32G per executor)

To mitigate this, one can set `spark.shuffle.spill.numElementsForceSpillThreshold` to force the spill on disk. While this config works, it is not flexible enough as it's expressed in number of elements, and in our case we run multiple shuffles in a single job and element size is different from one stage to another.


This patch extends the spill threshold behaviour and adds two new parameters to control the spill based on memory usage:

- `spark.shuffle.spill.map.maxRecordsSizeForSpillThreshold`
- `spark.shuffle.spill.reduce.maxRecordsSizeForSpillThreshold`


## How was this patch tested?

- internal e2e testing using large jobs
First, in our case for heavy RDD heavy shuffle jobs without setting the existing records based spill threshold (`spark.shuffle.spill.numElementsForceSpillThreshold`) the job in unstable and fails consistently with OOME in executors.
Trying to find the right value for `numElementsForceSpillThreshold` proved to be impossible.
Trying to maximize the job throughput (e.g. memory usage) while ensuring stability rendered to us an unbalanced usage across multiple stages of the job (in memory cached "elements" vary in size on map and reduce side combined with multiple map-reduce shuffles where "elements" are different)
Overall the best we could get in terms of memory usage is depicted in this snapshot:
![image](https://user-images.githubusercontent.com/728853/74935425-e526c080-53f0-11ea-8a2a-54c99ec2cef9.png)

Working from here, and using `spark.shuffle.spill.map.maxRecordsSizeForSpillThreshold` (map side, size-based spill) and `spark.shuffle.spill.numElementsForceSpillThreshold` (reduce side, size-based spill) we could maximize the memory usage (and in turn job runtime) while still keeping the job stable:
![image](https://user-images.githubusercontent.com/728853/74935955-e2789b00-53f1-11ea-9c6f-5e7025bdf468.png)




- Running existing unit-tests

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

holdenk commented:
Jenkins, ok to test.
@amuraru consider putting [WIP] back in the title if you still have outstanding TODOs.

dacort commented on core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java:
```diff
@@ -396,9 +406,14 @@ public void insertRecord(
       spill();
-    }
+    } else if (inMemRecordsSize >= maxRecordsSizeForSpillThreshold) {
+    logger.info("Spilling data because size of spilledRecords crossed the threshold " +
```
Indentation got changed up here through `growPointerArrayIfNecessary` on line 416 below. Needs to be re-indented.

amuraru replied:
fixed  - thanks

dacort commented:
Unit tests that need fixing/extending in spark-sql module:
- `UnsafeKVExternalSorterSuite`
- `ExternalAppendOnlyUnsafeRowArraySuite`
- `ExternalAppendOnlyUnsafeRowArrayBenchmark`

AmplabJenkins commented:
Can one of the admins verify this patch?

amuraru committed:
[SPARK-27734][CORE][SQL] Add memory based thresholds for shuffle spill

When running large shuffles (700TB input data, 200k map tasks, 50k reducers on a 300 nodes cluster) the job is regularly OOMing in map and reduce phase.

IIUC ShuffleExternalSorter (map side) and ExternalAppendOnlyMap and ExternalSorter (reduce side) are trying to max out the available execution memory. This in turn doesn't play nice with the Garbage Collector and executors are failing with OutOfMemoryError when the memory allocation from these in-memory structure is maxing out the available heap size (in our case we are running with 9 cores/executor, 32G per executor)

To mitigate this, one can set spark.shuffle.spill.numElementsForceSpillThreshold to force the spill on disk. While this config works, it is not flexible enough as it's expressed in number of elements, and in our case we run multiple shuffles in a single job and element size is different from one stage to another.

This patch extends the spill threshold behaviour and adds two new parameters to control the spill based on memory usage:

- spark.shuffle.spill.map.maxRecordsSizeForSpillThreshold
- spark.shuffle.spill.reduce.maxRecordsSizeForSpillThreshold

AmplabJenkins commented:
Can one of the admins verify this patch?

amuraru commented:
Removed the `WIP` tag - the PR is still valid IMO,
Can a committer have a look please?

amuraru commented:
/cc @dongjoon-hyun as a committer

dongjoon-hyun commented:
Thank you for pinging me, @amuraru .
What about `TODO: extend existing unit-tests`?
Do you have a plan to finish that, too?

dongjoon-hyun commented:
BTW, this PR is abandoned too long. Please resolve the conflict and see the Jenkins result.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

amuraru commented:
@dongjoon-hyun sorry for dropping the ball here. 
We have been running this patch in prod with very good results for 1yr+ now - it would be helpful to have it integrated here mainstream..

I rebased on top of `master` and fixed all conflicts.
Also on 
> What about TODO: extend existing unit-tests?

I updated all unit-tests but not sure if net new UT are required - the changes are covered well by existing UTs. Let me know what you think.

gatorsmile commented:
ok to test

gatorsmile commented:
add to whitelist

SparkQA commented:
**[Test build #119010 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/119010/testReport)** for PR 24618 at commit [`ab410fc`](https://github.com/apache/spark/commit/ab410fc48a039f161693ea81f8d15dbd3041d57f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/23757/
Test PASSed.

SparkQA commented:
**[Test build #119010 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/119010/testReport)** for PR 24618 at commit [`ab410fc`](https://github.com/apache/spark/commit/ab410fc48a039f161693ea81f8d15dbd3041d57f).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/119010/
Test FAILed.

Ngone51 commented:
retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/23777/
Test PASSed.

SparkQA commented:
**[Test build #119031 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/119031/testReport)** for PR 24618 at commit [`ab410fc`](https://github.com/apache/spark/commit/ab410fc48a039f161693ea81f8d15dbd3041d57f).

SparkQA commented:
**[Test build #119031 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/119031/testReport)** for PR 24618 at commit [`ab410fc`](https://github.com/apache/spark/commit/ab410fc48a039f161693ea81f8d15dbd3041d57f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/119031/
Test FAILed.

holdenk commented:
At quick glance the java codegen test failure would probably be unrelated, jenkins retest this please.

SparkQA commented:
**[Test build #119496 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/119496/testReport)** for PR 24618 at commit [`ab410fc`](https://github.com/apache/spark/commit/ab410fc48a039f161693ea81f8d15dbd3041d57f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/24226/
Test PASSed.

SparkQA commented:
**[Test build #119496 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/119496/testReport)** for PR 24618 at commit [`ab410fc`](https://github.com/apache/spark/commit/ab410fc48a039f161693ea81f8d15dbd3041d57f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/119496/
Test FAILed.

amuraru commented:
Looking into it

HyukjinKwon commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -998,6 +998,26 @@ package object config {
+        "until we reach some limitations, like the max page size limitation for the pointer " +
+        "array in the sorter.")
+      .bytesConf(ByteUnit.BYTE)
+      .createWithDefault(Long.MaxValue)
```
@amanomer, you can make this configuration optional via `createOptional` to represent no limit.

Ngone51 replied:
Just a reminder, we need to attach `version` info for the new configuration now. Just use `.version()`.

dongjoon-hyun commented:
cc @dbtsai

jiangxb1987 commented on core/src/main/java/org/apache/spark/shuffle/sort/ShuffleExternalSorter.java:
```diff
@@ -416,6 +428,7 @@ public void insertRecord(Object recordBase, long recordOffset, int length, int p
     Platform.copyMemory(recordBase, recordOffset, base, pageCursor, length);
     pageCursor += length;
     inMemSorter.insertRecord(recordAddress, partitionId);
+    inMemRecordsSize += length;
```
Should we also include the uaoSize?

cloud-fan replied:
+1, the `pageCursor` is also increased by uaoSize and length

jiangxb1987 commented on core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java:
```diff
@@ -396,8 +407,11 @@ public void insertRecord(
         numElementsForSpillThreshold);
       spill();
+    } else if (inMemRecordsSize >= maxRecordsSizeForSpillThreshold) {
+      logger.info("Spilling data because size of spilledRecords crossed the threshold " +
```
Should we also include the number of records and threshold here?

cloud-fan replied:
+1, let's follow https://github.com/apache/spark/pull/24618/files#diff-3eedc75de4787b842477138d8cc7f150R413

jiangxb1987 commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -998,6 +998,26 @@ package object config {
       .createWithDefault(Integer.MAX_VALUE)
 
+  private[spark] val SHUFFLE_SPILL_MAP_MAX_SIZE_FORCE_SPILL_THRESHOLD =
+    ConfigBuilder("spark.shuffle.spill.map.maxRecordsSizeForSpillThreshold")
```
What does the "map" mean inside this config name?

jiangxb1987 replied:
Why is it necessary to have different threshold between map task and reduce task?

cloud-fan replied:
I have the same question. What's the use case to separate them differently?

jiangxb1987 commented:
This PR looks pretty good, it would be great if we can add some new benchmark case to ensure it doesn't bring in performance regression if config values are properly chosen.

manuzhang commented:
@amuraru 
May I ask how would you set those thresholds in regard to `spark.executor.memory` ? Would, say, `0.8 * spark.executor.memory`  be a good candidate for those values ?

amuraru commented:
Ack @manuzhang - that makes sense

manuzhang commented:
@amuraru I forgot `spark.memory.fraction` and `spark.memory.storageFaction`. How will they play together with the new configurations ?

I've been testing Spark Adaptive Query Execution (AQE) recently, where contiguous shuffle partitions are coalesced to avoid too many small tasks. The problem is, IIUC, AQE makes coalesce decisions based on size of serialized map outputs. When data from multiple map tasks get deserialized into memory of one reduce task, it could easily blow up. I have to set extremely large `spark.executor.memory` to avoid being killed by YARN while wasting some resources. I think this patch is crucial for AQE to work steadily.

@cloud-fan @maryannxue @JkSelf

cloud-fan commented on core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java:
```diff
@@ -411,6 +425,7 @@ public void insertRecord(
     Platform.copyMemory(recordBase, recordOffset, base, pageCursor, length);
     pageCursor += length;
     inMemSorter.insertRecord(recordAddress, prefix, prefixIsNull);
+    inMemRecordsSize += length;
```
ditto, uaoSize

cloud-fan commented:
I'm not a big fan of having a static size limitation, can we follow the design of Spark memory management and make it more dynamic? e.g. these "memory consumers" should report its memory usage to spark memory manager and spill if the manager asks you to do so.

Can we make the `ShuffleExternalSorter` extends `MemoryConsumer`?

Ngone51 commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -998,6 +998,26 @@ package object config {
+        "until we reach some limitations, like the max page size limitation for the pointer " +
+        "array in the sorter.")
+      .bytesConf(ByteUnit.BYTE)
+      .createWithDefault(Long.MaxValue)
```
ditto.

Ngone51 commented on core/src/main/scala/org/apache/spark/util/collection/Spillable.scala:
```diff
@@ -81,7 +85,11 @@ private[spark] abstract class Spillable[C](taskMemoryManager: TaskMemoryManager)
-    if (elementsRead % 32 == 0 && currentMemory >= myMemoryThreshold) {
+    // Check number of elements or memory usage limits, whichever is hit first
+    if (_elementsRead > numElementsForceSpillThreshold
+      || currentMemory > maxSizeForceSpillThreshold) {
```
I just wonder if we need `maxSizeForceSpillThreshold` here since we've already have memory towards  control here?

cxzl25 commented:
Backported this PR in Spark3 version, fixed some compilation issues and SMJ codegen support. Verified in the production environment, the task time is shortened, the number of spill disks is reduced, there is a better chance to compress the shuffle data, and the size of the spill to disk is also significantly reduced.

### Current

<img width="1281" alt="image" src="https://github.com/user-attachments/assets/b6e172b8-0da8-4b60-b456-024880d0987e">

```
24/08/19 07:02:54,947 [Executor task launch worker for task 0.0 in stage 53.0 (TID 1393)] INFO ShuffleExternalSorter: Thread 126 spilling sort data of 62.0 MiB to disk (11490  times so far)
24/08/19 07:02:55,029 [Executor task launch worker for task 0.0 in stage 53.0 (TID 1393)] INFO ShuffleExternalSorter: Thread 126 spilling sort data of 62.0 MiB to disk (11491  times so far)
24/08/19 07:02:55,093 [Executor task launch worker for task 0.0 in stage 53.0 (TID 1393)] INFO ShuffleExternalSorter: Thread 126 spilling sort data of 62.0 MiB to disk (11492  times so far)
24/08/19 07:08:59,894 [Executor task launch worker for task 0.0 in stage 53.0 (TID 1393)] INFO Executor: Finished task 0.0 in stage 53.0 (TID 1393). 7409 bytes result sent to driver
```

### PR
<img width="1294" alt="image" src="https://github.com/user-attachments/assets/aedb83a4-c8a1-4ac9-a805-55ba44ebfc9e">

mridulm commented:
Can you create a new PR against master @cxzl25 ? We can evaluate it for inclusion in 4.0

cxzl25 commented:
> Can you create a new PR against master

https://github.com/apache/spark/pull/47856

Addressed some previous comments in new PR, and added tests.

1. inMemRecordsSize accumulates uaoSize and length
2. Remove separate configurations for map and reduce
3. Config added version
4. Support SMJ Codegen

Please help review, thanks in advance!
